docs/ql-libraries/dataflow/dataflow.md

```diff
@@ -0,0 +1,472 @@
+must be provided.
+
+First, two types, `DataFlowCall` and `DataFlowCallable`, must be defined. These
+should be aliases for whatever language-specific class represents calls and
+callables (a "callable" is intended as a broad term covering functions,
+methods, constructors, lambdas, etc.). The call-graph should be defined as a
+predicate:
+```ql
+DataFlowCallable viableCallable(DataFlowCall c)
+```
+
+In order to connect data-flow across calls, the 4 `Node` subclasses
+`ArgumentNode`, `ParameterNode`, `ReturnNode`, and `OutNode` are used.
+Flow into callables from arguments to parameters are matched up using an
+integer position, so these two classes must define:
+```ql
+ArgumentNode::argumentOf(DataFlowCall call, int pos)
+ParameterNode::isParameterOf(DataFlowCallable c, int pos)
+```
+It is typical to use `pos = -1` for an implicit `this`-parameter.
+
+For most languages return-flow is simpler and merely consists of matching up a
+`ReturnNode` with the data-flow node corresponding to the value of the call,
+represented as `OutNode`.  For this use-case we would define a singleton type
+`ReturnKind`, a trivial `ReturnNode::getKind()`, and `getAnOutNode` to relate
+calls and `OutNode`s:
+```ql
+private newtype TReturnKind = TNormalReturnKind()
+
+ReturnKind ReturnNode::getKind() { any() }
+
+OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) {
+  result = call.getNode() and
+  kind = TNormalReturnKind()
+}
+```
+
+For more complex use-cases when a language allows a callable to return multiple
+values, for example through `out` parameters in C#, the `ReturnKind` class can
+be defined and used to match up different kinds of `ReturnNode`s with the
+corresponding `OutNode`s.
+
+## Flow through global variables
+
+Flow through global variables are called jump-steps, since such flow steps
+essentially jump from one callable to another completely discarding call
+context.
+
+Adding support for this type of flow is done with the following predicate:
+```ql
+predicate jumpStep(Node node1, Node node2)
+```
+
+If global variables are common and certain databases have many reads and writes
+of the same global variable, then a direct step may have performance problems,
+since the straight-forward implementation is just a cartesian product of reads
+and writes for each global variable. In this case it can be beneficial to
+remove the cartesian product by introducing an intermediate `Node` for the
+value of each global variable.
+
+Note that, jump steps of course also can be used to implement other
+cross-callable flow. As an example Java also uses this mechanism for variable
+capture flow. But beware that this will lose the call context, so normal
+inter-procedural flow should use argument-parameter-, and return-outnode-flow
+as described above.
+
+## Field flow
+
+The library supports tracking flow through field stores and reads. In order to
+support this, a class `Content` and two predicates
+`storeStep(Node node1, Content f, PostUpdateNode node2)` and
```

Comment:
We no longer require that the target of a store is a post-update node.

Reply:
Does that mean that `storeIsPostUpdate` should be removed from `DataFlowImplConsistency.qll`?
Also, could that change be the reason that `storeIsPostUpdate` currently does not compile for me? (It complains that one branch of `Node` is incompatible with `PostUpdateNode`)

Reply:
Yes, I think `storeIsPostUpdate` should be removed.

Reply:
Hm, changing my `storeStep` so that the target _is_ a post-update node made the test compile (and pass).

Reply:
I'd say that the recommendation is that stores target postupdate nodes. I'm not sure I'm completely convinced by Toms use-case for it not to be the case.

Reply:
Reads are simple in this regard, they don't change the state of an object. Any flow step that changes the state of an object should target a `PostUpdateNode` - and store steps tend to do that.